In [1]:
%%HTML
<!-- some nicer latex formatting -->
<style>
.jp-MarkdownOutput {
    max-width: 40vw;
}
</style>

<script>
    MathJax.Hub.Config({
        displayAlign: "left",
        displayIndent: "2em",
        "HTML-CSS": {
            availableFonts: ["TeX"],
            preferredFont: "TeX",
            webfonts: "TeX"
        }
    });
    MathJax.Hub.Queue(
        ["resetEquationNumbers", MathJax.InputJax.TeX],
        ["PreProcess", MathJax.Hub],
        ["Reprocess", MathJax.Hub]
    );
</script>


# Information Entropy

This notebook is an introduction to information entropy. It tries to bring together various interpretations and uses of entropy, but largely from a machine learning perspective

## Definition

The definition of the entropy of a random variable is:
> the average level of "information", "surprise", or "uncertainty" inherent in the variable's possible outcomes</br>
> -- [Wikipedia](https://en.wikipedia.org/wiki/Entropy_(information_theory)#Definition)

This is a little ambiguous. Let's take a look at the formula. The average entropy of a random variable $X$ is given by

$$\mathcal{H}(X) = \mathbb{E}\left[ -\log X\right]$$

Well, this doesn't clarify too much either. Let's go back to first principles and build some intuition


## Information, Uncertainty, and Entropy

Suppose I have written down an integer, $X$, chosen uniformly at random between $1$ and $100$.

You don't know what the number is so you are missing some _information_. We want to measure the level of _**missing information**_ which we will call **entropy**. Equivalently, you could say there is a level of **uncertainty** in $X$. These are just both sides of the same coin and we will use these terms interchaneably.

Consider two different events:
* $A$: $X$ is between $1$ and $50$
* $B$: $X$ is between $1$ and $10$

If I told you that $B$ was true, then you would know more about the number than if I told you $A$ was true. Indeed, $B \subset A.$ So by knowing $B$ is true you know $A$ (plus a bit more). This means I must have provided _more_ information by sharing that $B$ was true. Equivalently, the level of uncertainty decreases more if I tell you $B$ is true.

On the other hand, if I told you that $B$ was not true, then you would only know that the number is between $11$ and $100$, and I would have provided _less_ information than sharing that $A$ was true.

Moreover, if you already knew the number was $5$, then $P(X=5)=1$, there would be no uncertainty, and telling you the number would convey _no_ information. 

If we look from a probability perspective then
* $P(A) = 0.5$
* $P(B) = 0.1$
* $P(\text{not}\, B) = 0.9$

Now we can make a key observation about Information Entropy:
> **Low probability events contain more information than high probability events.**

Or..
> **Low probability events have higher entropy than high probability events.**



## Defining a Measure of Information

We have some intuition that a probability event, $E$, represents a certain level of (missing) information $I(E)$. We haven't defined any units for information, so at the moment it is just an abstract quantity, and what matters is the relativity of $I$ between certain events. First we will try to posit some properties that we would like $I$ to have, and use them to define $I$.

### Existing Information

In the previous example there was a key underlying assumption: you knew that $X$ was uniformly distributed between $1$ and $100$. That is, you knew the sample space, the probability distribution and the event $A$. Therefore, the increase in information when event $A$ is realised does not contain information about the sample space itself (because it is already known). 

To put it another way, the uncertainty of event $A$ is the same as the level of uncertainty in a fair coin toss coming up heads. Both have probability $0.5$ and both in a sense "halve the sample space".

So even though the sample space of a fair coin toss, $Y$, only has $2$ elements, $\{H, T\}$, and the sample space of $X$ has $100$ elements, the _decrease_ in uncertainty / _increase_ in information is the same. The difference between these two events was already existing information.

Now we have some intuition about another property of entropy:
> **the entropy of an event $E$ does not depend on the sample space, only the probability $p(E)$**

That is, there is some function $I^*$ such that 

$$I(E) = I^*(p(E)) \tag{*}$$

### Linearity

Let's now continue to consider the random variable $Y$ representing a (not necessarily fair) coin toss, and the event $H_Y$ that the the coin comes up heads. If $Y$ comes up heads then we have learned $I(H_Y)$ units of information. If we flip another coin $Z$, which also comes up heads then we again learn $I(H_Z)$ units of information. The total information we learned is $I(H_Y) + I(H_Z)$. That is, when two events are independent

$$I(E_1 \land E_2) = I(E_1) + I(E_2)$$

However, these events are independent so

$$p(E_1 \land E_2) = p(E_1) p(E_2)$$

And combining with equation $(*)$

$$I^*(xy) = I^*(x) + I^*(y)$$

Where $x,y \in [0, 1]$. In fact, this is equivalent to [Cauchy's functional equation](https://en.wikipedia.org/wiki/Cauchy%27s_functional_equation) under the substitution $f(x)= g(e^x)$ and there is only one family of (continuous) solutions: 

$$I^*(x) = \alpha \log_e(x)$$

Where $\alpha \in \mathbb{R}$. So now we have a definition of $I$

$$I(E) = \alpha \log_e p(E) \tag{**}$$

### Some Observations

Notice that our definition of $I$ has a free parameter $\alpha$. This is because we have not defined any units for information.

Typically $\alpha$ is set to $-1$, which ensures that $I$ is always positive and conforms to the observation that low probability events have higher entropy.

If we set $\alpha = \dfrac{-1}{\log_e 2}$ then we change the basis: $I(E) = -\log_2 p(E)$, and we can make an interpretation of $I$ as the number of bits required to encode the information. This is also called [Shannon Entropy](https://en.wiktionary.org/wiki/Shannon_entropy) after Claude Shannon who introduced this idea. In particular the event of a fair coin coming up heads $I(H) = -\log_2 0.5 = 1$ bit. We will come back to this shortly.

As an aside (feel free to skip this part), setting $\alpha = k_B = 1.380649\cdot10^{−23}$ known as the [Boltzman Constant](https://en.wikipedia.org/wiki/Boltzmann_constant) provides a connection to the interpretation of entropy in physics. This interpretation deals with the configuration of molecules in a system. In essence, if there are some molecules in $W$ possible configurations (microstates) then the entropy is given by $S = k_B \log_e W$. Now, if each configuration is equally likely then the probability that they are in a specific microstate is $\frac{1}{W}$, and the entropy is... $-k_B \log_e \frac{1}W$. Which looks just like our interpretation of information entropy.

### Information in Bits

Information/Entropy can be described as

> the number of bits required to encode and transmit an event.

I have so far avoided this analogy because I think it can cause some confusion.

For instance, the event $HTHHT$ with respect to a fair coin being flipped 5 times, could be encoded $01001$. This provides some intuition that an event with probability $\frac1{2^5}$ has entropy $5$.

But what about a weighted coin that comes up heads with probability $0.9$ and tails $0.1$. A sequence of $5$ such coin tosses could still be "encoded" as $01001$, so we might feel that these have the same entropy. But we already know this event has entropy $I(HTHHT) = -\log_2(0.9^30.1^2) \approx 7.1$. _So what is the problem with this intuition?_

The problem is that this event has a lower probability than the fair coin equivalent, so it has higher entropy. A more explicit description might be:

> **entropy is the number of _units of information_ required to encode and transmit an event**

And recall from the previous section that we made this interpretation of encoding bits by setting $\alpha = \frac{-1}{\log_e 2}$. That is, _we_ specified that 1 bit is the amount of information transferred in the realisation of a _fair_ coin toss, not the othe way around!

We could just as easily define the realisation of a fair coin toss to be 2 bits or 3.14 bits. But by this definition we get the interpretation of information as a number of bits, like a computer.

# Entropy of a Random Variable

So far we have only considered single probabilistic events, but not random variables themselves. We have already seen that different realisations of a random variable $X$ can convey different levels of information (as with a weighted coin). So in fact, it doesn't make sense to talk about the information / level of uncertainty / entropy of a random variable $X$ itself, we can only talk about it's _expected entropy_. Or, as we saw in the first section:

$$\mathcal{H}(X) = \mathbb{E}\left[-\log X \right]$$ 

So for discrete distributions we have

$$\mathcal{H}(X) = -\sum_{j} p_j \log p_j$$

And in contradiction to what I just said, we will refer to $\mathcal{H}(X)$ as the entropy of $X$, even though it's really the expected entropy.

So the entropy of a fair coin toss, $Y$ is 

$$\mathcal{H}(Y) = -(0.5\log_2(0.5) + 0.5\log_2(0.5)) = \log_2(2) = 1$$

And the entropy of a weighted coin $Z$ that comes up heads with probability $0.9$ is

$$\mathcal{H}(Z) = -(0.9\log_2(0.9) + 0.1\log_2(0.1)) \approx 0.47$$

Notice that the weighted coin has lower entropy because there is less uncertainty (we are more confient that the coin will come up heads). We're now touching on the idea of maximum entropy. We'll come back to this later

## Back to Bits

Perhaps things still feel a bit abstract. Let's use our current understanding to present a more concrete example for a use of Entropy -- compression.

If we want to convert a string of text to a string of bits (i.e. real computer bits), we could use an encoding to first map each character to a representation in bits (not necessarily all the same length), and replace each character with it's binary representation. If we want to do this conversion in an optimal way (i.e. using the least number of bits), then more common characters should be assigned shorter binary representations.

The [source encoding theorem](https://en.wikipedia.org/wiki/Shannon%27s_source_coding_theorem) says that this compression will be optimal when each character's representation in bits has length $-\log_2 f$, where $f$ is the frequency of the character in the string.

Now, clearly you cannot use a fractional number of bits in practice, but the source encoding theorem gives the lower bound for the maximum compression, which is the entropy of the character distribution. Neat!

## Cross Entropy and Kullback-Leibler Divergence

Now let's go back to the initial example of the (discrete) uniform distribution $X \sim U(1, 100)$. For each integer $1 \le j \le 100$, the probability mass is $p_j = \frac{1}{100}$.

Suppose you had a different belief about the distribution of $X$. You assigned each integer a probability mass $q_j$. What would be the expected entropy that arises from using your probability mass $Q$ rather than the true probability mass $P$?

Well, each event $j$ still occurs with probability $p_j$, but you assigned a different level of entropy $-\log_2(q_j)$. The expectation would therefore be

$$\mathcal{H}(P, Q) = -\sum_j p_j \log_2(q_j)$$

This is referred to as the _cross entropy_ and, as it turns out, this will always overestimate the true entropy $\mathcal{H}(P)$. That is $\mathcal{H}(P, Q) \ge \mathcal{H}(P)$. The different between the cross entropy and the true entropy is the [Kullback-Leibler divergence](https://en.wikipedia.org/wiki/Kullback%E2%80%93Leibler_divergence), representing how much we have overestimated the entropy by using the incorrect distribution $Q$. That is

$$\mathcal{D}_{KL}(P\parallel Q) = \mathcal{H}(P, Q) - \mathcal{H}(P) = -\sum_j p_j \log_2 \frac{q_j}{p_j}$$

### KL Divergence is always positive

We just claimed that cross entropy always overestimates the true entropy. Equivallently the KL Divergence is always positive. Since the proof of this is short and sweet I'll include it here.

First note the taylor expansion of $\log_e$ when $|x| < 1$

$$
\begin{align}
-\log_e (x) &= -\log_e(1 - (1-x)) \\
           &= \sum_{n=1}^{\infty} \frac{(1-x)^n}{n!} \\
           &\ge (1-x)
\end{align}
$$

So

$$
\begin{align}
\mathcal{D}_{KL}(P\parallel Q) 
    &= -\sum_j p_j \log_2 \frac{q_j}{p_j} \\
    &\ge \sum_j p_j \left(1 - \frac{q_j}{p_j}\right) \\
    &= \sum_j (p_j - q_j) \\
    &= (1 - 1) = 0           
\end{align}
$$

With equality if and only if when $P = Q$


## Bits Revisited

When we first introduced the idea of entropy as the number of bits of information, we asked why encoding the flips of a weighted coin the same way as a fair coin gave us the wrong intuition about entropy. Well, now we can see that encoding the weighted coin like the fair coin, would be to assume the incorrect distribution of the coin.

That is, while the entropy of $n$ flips of the weighted coin, $P$, is

$$\mathcal{H}(P) = n(-0.9\log_2(0.9)-0.1\log_2(0.1)) \approx 0.47n$$

If we had assumed the incorrect distribution of a fair coin, $Q$,

$$\mathcal{H}(P, Q) = n(-0.9\log_2(0.5)-0.1\log_2(0.5)) = n$$

That is if we encode the $n$ flips with $n$ bits then we used $\mathcal{D}(P \parallel Q) \approx n - 0.47n = 0.53n$ too many bits. Or equivalently, we overestimated the entropy by $0.53n$


## Entropy and Machine Learning

Ok, this is great and all, but how does one apply the idea of cross entropy or KL divergence to ML? 

One place where cross entropy can be found is in the loss function of a classification model.

Suppose a model, $M$, is predicting one of $n$ classes $C_1, C_2,...,C_n$. For one observation $Y=C_k$ and data $X$, the model predicts $\hat{Y}=M(X)$, where $\hat{Y}$ is a vector of probability masses summing to $1$ and representing the weight that the model places on each of the $n$ classes. The _true distribution_ is the one-hot-encoding of $Y$ (meaning a vector with a $1$ in position $k$ and $0$ elsewhere), which we'll denote $Y^*$. The cross entropy is then calculated as:

$$\mathcal{H}(Y^*,\hat{Y}) = - \sum_j y^*_j \log \hat{y}_j = -\log y_k$$

Where it is understood that $p \log p = 0$ when $p=0$ because $x$ dominates $\log x$.

The idea is that the this is the loss function that you minimise while training the model. Note that because $Y^*$ is the distribution that says the class is definitely $C_k$, then there is no uncertainty and the entropy of $Y^*$ is $0$. So in this case the cross entropy is also the KL divergence.

Also note that because this all boils down to taking a single log, this is also called the log loss.

Man, that was a long way to come to discover that Cross Entropy loss is just log loss. Let's see a more interesting example.



## Information Gain and Decision Trees

> Mathematicans are low entropy. They take lots of abstract ideas, relate them and give it all a single name.
> Data Scientists are high entropy. They take one idea and give it 20 different names.

To that quip, Information Gain is basically just another name for KL Divergence used with reference to decision trees.

Suppose we are building a decision tree $T$, that predicts one of $n$ classes $C_1, C_2, ..., C_n$ from $m$ predictors. 

Our data is $N$ observations $y=C_k$, each with $m$ predictors $(x_1, x_2, ..., x_m)$, where each predictor $x_j$ is one of $X_j$ classes (e.g. the first predictor is sex: $x_1=\text{female}$ and $X_1 = \{\text{male, female} \}$).

To determine which predictor should split on at the next node of the decision tree, you select the one that decreases entropy the most (i.e. has the highest information gain).

To compute the information gain for class $X_j$, we take the difference between the entropy of all observations $\mathcal{H}(Y)$, and the weighted sum of the entropies over the $|X_j|$ different partitions of $Y$.

Let's do this one with an explicit example. We want to predict a child's sport preference based on their height and their sex.

In [2]:
import pandas as pd

# some contrived data
df = pd.DataFrame({
    "preference": ["soccer"]*10 + ["basketball"]*10,
    "height": ["short"]*7 + ["tall"]*13,
    "sex": ["male", "female"]*10,
})

df

preference height     sex
0       soccer  short    male
1       soccer  short  female
2       soccer  short    male
3       soccer  short  female
4       soccer  short    male
5       soccer  short  female
6       soccer  short    male
7       soccer   tall  female
8       soccer   tall    male
9       soccer   tall  female
10  basketball   tall    male
11  basketball   tall  female
12  basketball   tall    male
13  basketball   tall  female
14  basketball   tall    male
15  basketball   tall  female
16  basketball   tall    male
17  basketball   tall  female
18  basketball   tall    male
19  basketball   tall  female

If we just look at the data it seems evident that taller children like basketball, and sex has no impact at all. So we should expect that the information gain from splitting on height is high, and the gain from splitting on sex should be low.

In [3]:
import math

def H(df: pd.DataFrame, response: str) -> float:
    """ compute the entropy of the empirical distriution generated from
        the observations in column `response` of the DataFrame
    """
    probabilities = df[response].value_counts() / len(df)
    return sum(-p*math.log(p, 2) for p in probabilities)


def H_conditional(df: pd.DataFrame, response: str, attribute: str) -> float:
    """ compute the average entropy of the empirical distriutions generated 
        from the observations in column `response` split by `attribute`
    """
    N = len(df)
    return sum(
        (len(group) / N) * H(group, response)
        for _, group in df.groupby(attribute)
    )


def InformationGain(df: pd.DataFrame, response: str, attribute: str) -> float:
    """ compute the information gain from splitting on `attribute` """
    return H(df, response) - H_conditional(df, response, attribute)


print("IG Height:", InformationGain(df, "preference", "height"))
print("IG Gender:", InformationGain(df, "preference", "sex"))

IG Height: 0.49342260576014463
IG Gender: 0.0


We see that the splitting on height reduces entropy by $0.49$ and splitting on sex does not reduce entropy at all, which is in line with out expectation. Note that because an equal number of boys and girls both like soccer and basketball, you do not gain any information about their sport preference if you know the sex of the child. That is, the information gain is zero.

In our decision tree, we would therefore split on height because it provided the largest decrease in entropy / decrease in uncertainty / KL divergence / information gain, or however you like to say it.

## Maximum Entropy

This tutorial wouldn't be complete if we didn't touch on the concept of maximum entropy. We noted previously that the information/entropy associated with a single event $E$ does not depend on the sample space. While this is true for single events, it is not true when we talk about the average entropy of a random variable.

When there are more possible outcomes, the average uncertainty (i.e. entropy) is higher 

Indeed, consider the entropy of $U(1, n)$,

$$\mathcal{H}(U(1,n)) = -\sum_{j=1}^n \frac{1}n \log_2\frac{1}n = \log_2 n$$

Which will tend to infinity as $n \to \infty$.

When we talk about maximum entropy we do so in relation to some constraint. 

For instance, when we restrict the sample space to have cardinality $n$ then the distribution with maxmimum entropy is the uniform distribution. That is $\mathcal{H}(X)$ is maximised when all the terms $p_j$ are equal.

This also coincides with the physics interpretation of maximum entropy in which particles are very dispersed (uniformly distributed) through space.

Now, if were talking about a continuous distribution and we knew the mean and variance, then the maximum entropy distribution would be the Gaussian!

Which brings us to our final point

## Continuous Distributions

In this tutorial we have only considered discrete distributions. How do we extend entropy to continuous ones? We would hope that we can simply write

$$\mathcal{H}(X) = -\int_{\mathbb{R}} f_X(x) \log f_X(x)\, dx$$

Unfortunately most of our intuition breaks down when we try to replace probability masses with probability densities. For one, a probability density can be greater than $1$, which would give us a negative entropy at that point. At least from an Machine Learning perspective, this is one reason why we see entropy as a loss function in classification problems and generally not in regression problems.

As it turns out, with some careful consideration, the continuous case can be made to work. But I will not go into that here because this tutorial is already quite long and has hopefully acheived my goal of providing an intuition for Entropy, with a little bit of math and rigour.